docs/Building SpiderMonkey.md

```diff
@@ -0,0 +1,62 @@
+# Building SpiderMonkey for Embedders #
+
+Use these instructions to build your own copy of SpiderMonkey.
+
+## Prerequisites ##
+
+You will need a **C++ compiler** that can handle the C++14 standard,
+**GNU Make**, **zlib**, and **libffi**.
+These can usually be installed with a package manager.
+You will also need **ICU** of at least version 59.1.
+ICU can also usually be installed with a package manager, but the
+version might not be new enough.
+
+> **NOTE** If you can't install a new enough version of ICU, you can
+> build the copy bundled with SpiderMonkey by omitting
+> `--with-system-icu` in the build instructions below, at the cost of a
+> longer build time.
+
+## Getting the source code ##
+
+Currently, the most reliable way to get the SpiderMonkey source code is
+to download the Firefox source.
+At the time of writing, the latest source for Firefox ESR 60, which
+contains the source for SpiderMonkey ESR 60, can be found here:
+http://ftp.mozilla.org/pub/firefox/releases/60.3.0esr/source/
+
+> **NOTE** Mozilla may be able to provide separate source packages for
+> SpiderMonkey in the future, but this is difficult for a number of
+> reasons.
+
+## Building SpiderMonkey ##
+
+First you should decide where you want to install SpiderMonkey.
+By default, it will install into `/usr/local`.
+You might want to pick some other location if `/usr/local` is not
+writable to you without superuser permissions, for example.
+
+```sh
+cd js/src
+mkdir _build
+cd _build
+../configure --disable-jemalloc  --enable-posix-nspr-emulation \
+    --with-system-zlib --with-system-icu --with-intl-api \
+    --enable-debug --enable-optimize AUTOCONF=autoconf
+make
+make install  # sudo if necessary
+```
+
+Add `--prefix=/my/prefix` (where `/my/prefix` is your chosen prefix) to
+the `configure` line if you chose a different prefix.
+
+If you are building a package for production, omit the `--enable-debug`.
+
+If you picked a different prefix to install into, and that prefix is not
+a standard place where libraries are loaded from, you may need to
+execute the following when you want to use the SpiderMonkey libraries,
+for example when building the examples from this repository.
+
```

Comment:
I'd stress the jemalloc problem again since it always surprises people:
(Feel free to rewrite entirely)

### Disabling jemalloc
One important configuration when getting started is the '--disable-jemalloc' flag. This will cause SpiderMonkey to use the default system allocator functions instead of a custom build of jemalloc. The default configuration is intended for a browser environment and requires linking the final application with a matching version of a library called mozglue. If one accidentally builds SpiderMonkey for their embedding without including the '--disable-jemalloc' flag, they usually quickly encounter strange crashes as items allocated in jemalloc allocator are freed on system allocator.